### PR TITLE
Independent Table Elements

### PR DESCRIPTION
**Pull request by aaroncox**
This PR breaks the dependencies between Contract, Table, and TableCursor - which allows developers the opportunity to use each element individually if they choose to do so.

### CODE REVIEW

File: src/contract/table-cursor.ts [hunk 1]
```diff
@@ -63,113 +55,65 @@ export class TableCursor<TableRow> {
...
 
-    /**
-     * Returns all rows in the cursor query.
-     *
-     * @returns All rows in the cursor query.
-     */
-    async all() {
-        const rows: TableRow[] = []
+    async all(): Promise<RowType[]> {
+        const rows: any[] = []
```
Comment (dafuga) [inline src/contract/table-cursor.ts:113]: This should be `const rows: RowType[] = []`, no?

Comment (dafuga) [reply src/contract/table-cursor.ts:113]:
```suggestion
        const rows: RowType[] = []
```

File: src/contract/table-cursor.ts [hunk 2]
```diff
@@ -63,113 +55,65 @@ export class TableCursor<TableRow> {
...
         }
         return rows
     }
-
-    /**
-     * Returns a new cursor with updated parameters.
-     *
-     * @returns A new cursor with updated parameters.
-     */
-    query(query: Query) {
```
Comment (dafuga) [inline src/contract/table-cursor.ts:163]: Hmm I guess one side effect of decoupling table and table cursor is that this now feels out of place 🤔

File: src/contract/table-cursor.ts [hunk 3]
```diff
@@ -63,113 +55,65 @@ export class TableCursor<TableRow> {
...
+        let lower_bound = this.params.lower_bound
+        const upper_bound = this.params.upper_bound
 
         if (this.next_key) {
             lower_bound = this.next_key
         }
 
-        let indexPosition = this.tableParams.index_position || 'primary'
-
-        if (this.indexPositionField) {
```
Comment (dafuga) [inline src/contract/table-cursor.ts:92]: I have to admit that I do love the fact that this is no longer needed 😁

File: src/contract/table-cursor.ts [hunk 4]
```diff
@@ -63,113 +55,65 @@ export class TableCursor<TableRow> {
...
-            index_position: indexPosition,
-            json: false,
-        })
+        }
 
-        let {rows} = result
-        this.next_key = result.next_key
+        const result = await this.client!.v1.chain.get_table_rows(query)
 
+        const rows =
```
Comment (dafuga) [inline src/contract/table-cursor.ts:85]: This statement variable assignment does a lot and is a bit difficult to follow so I would probably break it up into different steps.

File: src/contract/table-cursor.ts [hunk 5]
```diff
@@ -1,60 +1,52 @@
...
+
     private next_key: API.v1.TableIndexType | string | undefined
-    private tableParams: API.v1.GetTableRowsParams
+
     private endReached = false
-    private indexPositionField?: string
     private rowsCount = 0
     private maxRows: number = Number.MAX_SAFE_INTEGER
 
-    /**
```
Comment (dafuga) [inline src/contract/table-cursor.ts:28]: I know that these were far from perfect, but we will still want to have some JSDoc statements here, right?

File: src/contract/table-cursor.ts (hunk 1 above)
Comment (aaroncox) [reply src/contract/table-cursor.ts]: You're absolutely right.

Also, whatever this "Commit Suggestion" thing is, is awesome 😍 

It just let me click a button and merge that change in!

File: test/tests/cursor.ts [hunk 6]
```diff
@@ -0,0 +1,397 @@
+import {assert} from 'chai'
```
Comment (dafuga) [inline test/tests/cursor.ts:1]: 💯

File: test/tests/cursor.ts [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+import ContractKit, {Contract, TableCursor} from '$lib'
+
+import {Name, Serializer, UInt64} from '@greymass/eosio'
+import {makeClient} from '@wharfkit/mock-data'
+import {TrendingRow} from '$test/utils/decentium'
+import {DelegatedBandwidth, NameBid} from '$test/utils/eosio'
+
+const mockClient = makeClient('https://eos.greymass.com')
+
+suite('Cursor', () => {
```
Comment (dafuga) [inline test/tests/cursor.ts:12]: 💯

File: src/contract/table-cursor.ts (hunk 2 above)
Comment (aaroncox) [reply src/contract/table-cursor.ts:163]: Yeah if we want it back - it's going to need to be a function that translates a `TableCursor` back into a `Table` call, and that kinda re-adds that dependency this was seeking to remove. 

I'm thinking maybe we add this feature into whatever a future `QueryBuilder` might be, and leave it off for now.

File: test/tests/table.ts [hunk 8]
```diff
@@ -117,13 +146,14 @@ suite('Table', () => {
...
     suite('query', () => {
-        test('should allow you to chain index query statements', async () => {
-            const tableCursor = decentiumTrendingTable.query({from: 5, to: 10}).query({to: 8})
-            assert.deepEqual(
-                Serializer.objectify(await tableCursor.all()).map((row) => row.id),
-                [5, 6, 7, 8]
-            )
-        })
+        // TODO: Discuss and remove? No longer works since query is a Table method
+        // test('should allow you to chain index query statements', async () => {
```
Comment (dafuga) [inline test/tests/table.ts:150]: Yeah, I guess we can just remove this. I personally kind of liked the fact that you could chain query statements like this, but given that the cursor and table classes are now built to be used independently, it no longer really makes sense to have a query method on the cursor.

File: test/utils/decentium.ts [hunk 9]
```diff
@@ -0,0 +1,382 @@
+// generated by @greymass/abi2core
```
Comment (dafuga) [inline test/utils/decentium.ts:1]: Did you mean to commit this? 😄

File: test/utils/eosio.ts [hunk 10]
```diff
@@ -0,0 +1,799 @@
+// generated by @greymass/abi2core
```
Comment (dafuga) [inline test/utils/eosio.ts:1]: And this?

**Review by dafuga — APPROVED** [review]
The more I think about it, the more I realize that this is probably a good change to make! I left you some comments about some minor stuff, but overall the PR looks good! Thanks for working on this!

File: src/contract/table-cursor.ts (hunk 2 above)
Comment (dafuga) [reply src/contract/table-cursor.ts:163]: Sounds good!

File: src/contract/table-cursor.ts (hunk 1 above)
Comment (dafuga) [reply src/contract/table-cursor.ts]: Hehe yeah, it's pretty cool 😄

File: src/contract/table-cursor.ts (hunk 4 above)
Comment (aaroncox) [reply src/contract/table-cursor.ts]: I gave this a shot, and added some comments. 

https://github.com/wharfkit/contract/blob/ede31bfd05b00139969f82b06968d98b204b2f12/src/contract/table-cursor.ts#L85-L100

I'm not sure how much more I think we should split it out, unless we wanted to get rid of the ternary operator, and instead do... 

```ts
let rows = thing 
if (decode) { 
  rows = otherThing 
}
```

File: src/contract/table-cursor.ts (hunk 5 above)
Comment (aaroncox) [reply src/contract/table-cursor.ts:28]: Yeah absolutely. I just kept seeing so many things changing - that I wanted to redo them all if this was the direction we wanted to go 👍

File: test/tests/table.ts (hunk 8 above)
Comment (aaroncox) [reply test/tests/table.ts:150]: Removed for now, and left a link to this discussion in the commit.

File: test/utils/decentium.ts (hunk 9 above)
Comment (aaroncox) [reply test/utils/decentium.ts:1]: Yeah the unit tests are using it to test with `Struct` instances from abi2core.

**Pull request merged**